server/aesdsocket.c

```diff
@@ -163,17 +423,405 @@ static int daemon_init()
+        memcpy(rx_storage_buffer+(rx_storage_buffer_len), rx_buffer, (i));
+        rx_storage_buffer_len += (i);
+    }
+
+    // Will store into temp and then store after sanity check so that the original pointer is not lost
+    char* temp = (char*) realloc(rx_storage_buffer, rx_storage_buffer_len+1);
+    if(temp == NULL)
+    {
+        printf("\nError: Failed realloc(). Error code: %d\n", errno);
+        // Syslog the error into the syslog file in /var/log
+        syslog(LOG_ERR, "Error: Failed realloc(). Error code: %d", errno);
+        free(rx_storage_buffer);
+        // return -1;
+    }
+    else
+        rx_storage_buffer = temp;
+
+    *(rx_storage_buffer+(rx_storage_buffer_len)) = '\n';
+    rx_storage_buffer_len++;
+
+#if DEBUG
+    printf("Data Len Received: %d\n", rx_storage_buffer_len);
+    for(i=0;i<rx_storage_buffer_len;i++)
+    {
+        printf("%c", *(rx_storage_buffer+i));
+    }
+    printf("\n"); // It seems line the system was buffering the printf so absense of new line was making it buffer and printing only the next time new line was met
+    // Refer to https://stackoverflow.com/questions/39180642/why-does-printf-not-produce-any-output
+#endif
+
+    // The mutex object referenced by mutex shall be locked by a call to
+    // pthread_mutex_lock() that returns zero or [EOWNERDEAD].  If the
+    // mutex is already locked by another thread, the calling thread
+    // shall block until the mutex becomes available.
+    // int pthread_mutex_lock(pthread_mutex_t *mutex);
+    //  manpage: https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html
+    int status = pthread_mutex_lock(&mutex);
+    if(status != 0) // returns non zero on error
+    {
+        printf("\nError: Failed pthread_mutex_lock(). Error code: %d\n", errno);
+        // Syslog the error into the syslog file in /var/log
+        syslog(LOG_ERR, "Error: Failed pthread_mutex_lock(). Error code: %d", errno);
+        // return -1;
+        goto err_handle;
+    }
+
+    /*********************************************************************************************************
+                                    Writing to file /var/tmp/aesdsocketdata
+    **********************************************************************************************************/
+    // write() writes up to count bytes from the buffer starting at buf
+    // to the file referred to by the file descriptor fd.
+    // ssize_t write(int fd, const void *buf, size_t count);
+    // manpage: https://man7.org/linux/man-pages/man2/write.2.html
+    int bytes_written = write(socket_file_fd, rx_storage_buffer, rx_storage_buffer_len);
+    if(bytes_written == -1) // returns -1 on error else number of bytes written
+    {
+        printf("\nError: Failed write(). Error code: %d\n", errno);
+        // Syslog the error into the syslog file in /var/log
+        syslog(LOG_ERR, "Error: Failed write(). Error code: %d", errno);
+        // return -1;
+        goto err_handle;
```

Comment:
@nihalthirunakarasu you are missing mutex unlock in these and other error cases which might cause problems if multiple threads are running.